alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -1,47 +1,331 @@
 #pragma once
 
 #include "BlackboardBlueprint.h"
+#include "engine/Types.h"
+#include "engine/modelmanagement/Strings.h"
 #include <any>
+#include <array>
+#include <cassert>
+#include <exception>
 #include <mutex>
 #include <shared_mutex>
 #include <string>
 #include <type_traits>
 #include <unordered_map>
+#include <variant>
+#include <yaml-cpp/yaml.h>
 
 namespace alica
 {
+
+class BlackboardUtil;
+
+struct BlackboardException : public std::logic_error
+{
+    static constexpr const char* BB_EXCEPTION_PREFIX = "Blackboard exception: ";
+    explicit BlackboardException(const std::string& errMsg)
+            : std::logic_error(BB_EXCEPTION_PREFIX + errMsg)
+    {
+    }
+    BlackboardException(const BlackboardException&) = default;
+    BlackboardException& operator=(const BlackboardException&) = default;
+};
+
 class BlackboardImpl
 {
-public:
-    // Not for API use, but public to allow modifying without mutex when we know the behavior/plan is not running
+    using BBValueType = std::variant<std::monostate, bool, int64_t, uint64_t, double, std::string, std::any>;
+    static constexpr std::size_t BB_VALUE_TYPE_ANY_INDEX = 6;
+
     template <class... Args>
-    void registerValue(const std::string& key, Args&&... args)
+    static std::string stringify(Args&&... args)
     {
-        vals.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(std::forward<decltype(args)>(args)...));
+        std::ostringstream oss;
+        (oss << ... << std::forward<Args>(args));
+        return oss.str();
+    }
+
+public:
+    BlackboardImpl() = default;
+    BlackboardImpl(const BlackboardImpl&) = delete;
+    BlackboardImpl(const BlackboardBlueprint* blueprint)
```

Comment:
I think parsing the yaml etc should be done in the blueprint, not when constructing the blackboard

Reply:
ok, I will create another task for this & see how it can be done because the parsing related code is currently tied up in the blackboard class. Wrike task: https://www.wrike.com/open.htm?id=1033598893